R/ensemble.R

```diff
@@ -0,0 +1,175 @@
+#' Estimating Ensemble Kernel Matrices
+#' 
+#' Give a list of estimated kernel matrices and their weights.
+#' 
+#' There are three ensemble strategies available here:
+#' 
+#' \bold{Empirical Risk Minimization}
+#' 
+#' After obtaining the estimated errors \eqn{\{\hat{\epsilon}_d\}_{d=1}^D}, we
+#' estimate the ensemble weights \eqn{u=\{u_d\}_{d=1}^D} such that it minimizes
+#' the overall error \deqn{\hat{u}=\underset{u \in \Delta}{argmin}\parallel
+#' \sum_{d=1}^Du_d\hat{\epsilon}_d\parallel^2 \quad where\; \Delta=\{u | u \geq
+#' 0, \parallel u \parallel_1=1\}} Then produce the final ensemble prediction:
+#' \deqn{\hat{h}=\sum_{d=1}^D \hat{u}_d h_d=\sum_{d=1}^D \hat{u}_d
+#' A_{d,\hat{\lambda}_d}y=\hat{A}y} where \eqn{\hat{A}=\sum_{d=1}^D \hat{u}_d
+#' A_{d,\hat{\lambda}_d}} is the ensemble matrix.
+#' 
+#' \bold{Simple Averaging}
+#' 
+#' Motivated by existing literature in omnibus kernel, we propose another way
+#' to obtain the ensemble matrix by simply choosing unsupervised weights
+#' \eqn{u_d=1/D} for \eqn{d=1,2,...D}.
+#' 
+#' \bold{Exponential Weighting}
+#' 
+#' Additionally, another scholar gives a new strategy to calculate weights
+#' based on the estimated errors \eqn{\{\hat{\epsilon}_d\}_{d=1}^D}.
+#' \deqn{u_d(\beta)=\frac{exp(-\parallel \hat{\epsilon}_d
+#' \parallel_2^2/\beta)}{\sum_{d=1}^Dexp(-\parallel \hat{\epsilon}_d
+#' \parallel_2^2/\beta)}}
+#' 
+#' @param n A numeric number specifying the number of observations.
+#' @param D A numeric number specifying the number of kernels in the kernel
+#' library.
+#' @param strategy A character string indicating which ensemble strategy is to
+#' be used.
+#' @param beta A numeric value specifying the parameter when strategy = "exp".
+#' @param error_mat A n*D matrix indicating errors.
+#' @param A_hat A list of projection matrices for every kernels in the kernel
+#' library.
+#' @return \item{A_est}{A list of estimated kernel matrices.}
+#' 
+#' \item{u_hat}{A vector of weights of the kernels in the library.}
+#' @author Wenying Deng
+#' @seealso mode: \code{\link{tuning}}
+#' @references Jeremiah Zhe Liu and Brent Coull. Robust Hypothesis Test for
+#' Nonlinear Effect with Gaus- sian Processes. October 2017.
+#' 
+#' Xiang Zhan, Anna Plantinga, Ni Zhao, and Michael C. Wu. A fast small-sample
+#' kernel inde- pendence test for microbiome community-level association
+#' analysis. December 2017.
+#' 
+#' Arnak S. Dalalyan and Alexandre B. Tsybakov. Aggregation by Exponential
+#' Weighting and Sharp Oracle Inequalities. In Learning Theory, Lecture Notes
+#' in Computer Science, pages 97– 111. Springer, Berlin, Heidelberg, June 2007.
+#' @examples
+#' 
+#' 
+#' ##ensemble(n = 50, D = 6, strategy = "erm", beta = 1, error_mat, A_hat)
+#' 
+#' 
+#' @export ensemble
+ensemble <-
+  function(n, D, strategy, beta, error_mat, A_hat){
+
+    if (strategy == "erm"){
```

Comment:
Its a good strategy to write each strategy as separate functions, then call these functions using do.call().

for example, define three functions:

```
ensemble_erm <- function(n, D, error_mat, A_hat) {...}
ensemble_avg <- function(n, D, error_mat, A_hat) {...}
ensemble_exp <- function(n, D, error_mat, A_hat) {...}

```
then in the main function:

```
ensemble <-
  function(n, D, strategy, beta, error_mat, A_hat){
    strategy <- match.arg(kernel,c("avg","exp","erm"))
    func_name <- paste0("ensemble_", strategy)
    do.call(func_name, list(n = n, D = D, error_mat = error_mat, A_hat = A_hat))
}
```